### ISSUE TITLE
SONiC Exporter not working on Sonic Open Source Version

### ISSUE THREAD

**Issue by rokernel** [kamelnetworks/sonic_exporter#3]
Hello,

Sonic exporter is not working on Sonic Open Source Version. Is is supposed to work with this or just on Enterprise version?

It's crashing when trying to access the "http://localhost:port/metrics"

I have tried with the last 202211 Broadcom build.

```
admin@sonic:~$ sudo sonic-package-manager install --from-tarball sonic_exporter_v0.2.0.tar.gz --enable
sonic_exporter_v0.2.0.tar.gz is going to be installed, continue? [y/N]: y

Loading layer de1c39ee943b 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████|        0.00/0.00 Mb [00:00<00:00, 0.00 Mb/s]
Loading layer db68f3f0ad09 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████|         17.85/17.85 Mb [00:00<00:00, 59.64 Mb/s]
Loading layer 48fe1618bac3 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████|         0.00/0.00 Mb [00:00<00:00, 0.33 Mb/s]
generated /usr/bin/sonic_exporter.sh
generated /usr/local/bin/sonic_exporter.sh
generated /usr/lib/systemd/system/sonic_exporter.service
sonic_exporter entry is added to AUTO_TECHSUPPORT_FEATURE table

admin@sonic:~$ sudo config sonic-exporter
Usage: config sonic-exporter [OPTIONS] COMMAND [ARGS]...

  Configure Prometheus exporter for SONiC

Options:
  -h, -?, --help  Show this message and exit.

Commands:
  port  Set the port that the exporter is listening to.
  vrf   Set the VRF that the exporter is listening inside.

admin@sonic:~$ sudo config sonic-exporter port 9102

admin@sonic:~$ curl http://localhost:9102
curl: (7) Failed to connect to localhost port 9102: Connection refused
admin@sonic:~$ curl http://localhost:9102/metrics
curl: (7) Failed to connect to localhost port 9102: Connection refused

admin@sonic:~$ sudo docker ps
CONTAINER ID   IMAGE                                COMMAND                  CREATED        STATUS        PORTS     NAMES
566d6dc6904a   docker-sonic-telemetry:latest        "/usr/local/bin/supe…"   23 hours ago   Up 23 hours             telemetry
db5e0070dd0f   docker-sonic-mgmt-framework:latest   "/usr/local/bin/supe…"   23 hours ago   Up 23 hours             mgmt-framework
73d15696bfad   docker-snmp:latest                   "/usr/local/bin/supe…"   23 hours ago   Up 23 hours             snmp
3aaba76cb427   docker-teamd:latest                  "/usr/local/bin/supe…"   23 hours ago   Up 23 hours             teamd
774c9687f4c8   docker-fpm-frr:latest                "/usr/bin/docker_ini…"   23 hours ago   Up 23 hours             bgp
8748e8154d31   docker-lldp:latest                   "/usr/bin/docker-lld…"   23 hours ago   Up 23 hours             lldp
0626f850a892   docker-platform-monitor:latest       "/usr/bin/docker_ini…"   23 hours ago   Up 23 hours             pmon
7d3eaaf7e2a0   docker-router-advertiser:latest      "/usr/bin/docker-ini…"   23 hours ago   Up 23 hours             radv
facdfb328b7d   docker-syncd-brcm:latest             "/usr/local/bin/supe…"   23 hours ago   Up 23 hours             syncd
09845b59068c   docker-orchagent:latest              "/usr/bin/docker-ini…"   23 hours ago   Up 23 hours             swss
ad1f6154685c   docker-eventd:latest                 "/usr/local/bin/supe…"   23 hours ago   Up 23 hours             eventd
db6e0fdce9a6   docker-database:latest               "/usr/local/bin/dock…"   23 hours ago   Up 23 hours             database

admin@sonic:~$ sudo systemctl status sonic
sonic-delayed.target       sonic_exporter.service     sonic-hostservice.service  sonic.target
admin@sonic:~$ sudo systemctl status sonic_exporter.service
● sonic_exporter.service - sonic_exporter container
     Loaded: loaded (/lib/systemd/system/sonic_exporter.service; enabled; vendor preset: enabled)
    Drop-In: /etc/systemd/system/sonic_exporter.service.d
             └─auto_restart.conf
     Active: active (running) since Thu 2022-12-01 20:23:21 UTC; 12s ago
    Process: 1016492 ExecStartPre=/usr/local/bin/sonic_exporter.sh start (code=exited, status=0/SUCCESS)
   Main PID: 1016594 (sonic_exporter.)
      Tasks: 3 (limit: 19107)
     Memory: 18.9M
     CGroup: /system.slice/sonic_exporter.service
             ├─1016594 /bin/bash /usr/local/bin/sonic_exporter.sh wait
             ├─1016605 /bin/bash /usr/bin/sonic_exporter.sh wait
             └─1016606 python3 /usr/local/bin/container wait sonic_exporter

Dec 01 20:23:21 sonic container[1016523]: container_start: sonic_exporter: set_owner:local fallback:True remote_state:none server_connected:false
Dec 01 20:23:21 sonic container[1016523]: docker cmd: start for sonic_exporter
Dec 01 20:23:21 sonic container[1016523]: container_start: END
Dec 01 20:23:21 sonic systemd[1]: Started sonic_exporter container.
Dec 01 20:23:21 sonic root[1016602]: Warm boot flag: sonic_exporter false.
Dec 01 20:23:21 sonic container[1016606]: container_wait: BEGIN
Dec 01 20:23:21 sonic container[1016606]: read_data: config:True feature:sonic_exporter fields:[('set_owner', 'local'), ('no_fallback_to_local', False), ('state', 'disabled')] val:['local', False, 'enabled']
Dec 01 20:23:21 sonic container[1016606]: read_data: config:False feature:sonic_exporter fields:[('current_owner', 'none'), ('remote_state', 'none'), ('container_id', '')] val:['none', 'none', '']
Dec 01 20:23:21 sonic container[1016606]: container_wait: sonic_exporter: set_owner:local ct_owner:none state:none id:sonic_exporter pend=0
Dec 01 20:23:21 sonic container[1016606]: container_wait: END -- transitioning to docker wait

admin@sonic:~$ sudo docker logs -f sonic_exporter
level=info msg="Starting up" exporter-git-hash=e1d7f32 exporter-version=0.2.0 lib-node-version=v1.3.1 lib-prometheus-version=v1.11.0
level=info msg="Configuration loaded" port=1234 vrf=
level=info msg="node-library: Parsed flag --collector.filesystem.mount-points-exclude" flag="^/(dev|proc|run/credentials/.+|sys|var/lib/docker/.+)($|/)"
level=info msg="node-library: Parsed flag --collector.filesystem.fs-types-exclude" flag="^(autofs|binfmt_misc|bpf|cgroup2?|configfs|debugfs|devpts|devtmpfs|fusectl|hugetlbfs|iso9660|mqueue|nsfs|overlay|proc|procfs|pstore|rpc_pipefs|securityfs|selinuxfs|squashfs|sysfs|tracefs)$"
level=info msg="node-library: Parsed flag --collector.systemd.unit-include" flag=.+
level=info msg="node-library: Parsed flag --collector.systemd.unit-exclude" flag=".+\\.(automount|device|mount|scope|slice)"
level=warning msg="node-library: Unable to get systemd dbus connection, defaulting systemd version to 0" err="dial unix /run/systemd/private: connect: no such file or directory"
level=warning msg="node-library: Detected systemd version is lower than minimum" current=0 minimum=212
level=warning msg="node-library: Some systemd state and timer metrics will not be available"
level=info msg="SONiC Prometheus exporter running"
level=info msg="Configuration change detected, restarting exporter..."
level=info msg="Starting up" exporter-git-hash=e1d7f32 exporter-version=0.2.0 lib-node-version=v1.3.1 lib-prometheus-version=v1.11.0
level=info msg="Configuration loaded" port=9102 vrf=
level=info msg="node-library: Parsed flag --collector.filesystem.mount-points-exclude" flag="^/(dev|proc|run/credentials/.+|sys|var/lib/docker/.+)($|/)"
level=info msg="node-library: Parsed flag --collector.filesystem.fs-types-exclude" flag="^(autofs|binfmt_misc|bpf|cgroup2?|configfs|debugfs|devpts|devtmpfs|fusectl|hugetlbfs|iso9660|mqueue|nsfs|overlay|proc|procfs|pstore|rpc_pipefs|securityfs|selinuxfs|squashfs|sysfs|tracefs)$"
level=info msg="node-library: Parsed flag --collector.systemd.unit-include" flag=.+
level=info msg="node-library: Parsed flag --collector.systemd.unit-exclude" flag=".+\\.(automount|device|mount|scope|slice)"
level=warning msg="node-library: Unable to get systemd dbus connection, defaulting systemd version to 0" err="dial unix /run/systemd/private: connect: no such file or directory"
level=warning msg="node-library: Detected systemd version is lower than minimum" current=0 minimum=212
level=warning msg="node-library: Some systemd state and timer metrics will not be available"
level=info msg="SONiC Prometheus exporter running"
^C

admin@sonic:~$ sudo docker ps
CONTAINER ID   IMAGE                                COMMAND                  CREATED              STATUS          PORTS     NAMES
ee0d270105f3   c2ad8968cced                         "./sonic_exporter"       About a minute ago   Up 42 seconds             sonic_exporter
566d6dc6904a   docker-sonic-telemetry:latest        "/usr/local/bin/supe…"   23 hours ago         Up 23 hours               telemetry
db5e0070dd0f   docker-sonic-mgmt-framework:latest   "/usr/local/bin/supe…"   23 hours ago         Up 23 hours               mgmt-framework
73d15696bfad   docker-snmp:latest                   "/usr/local/bin/supe…"   23 hours ago         Up 23 hours               snmp
3aaba76cb427   docker-teamd:latest                  "/usr/local/bin/supe…"   23 hours ago         Up 23 hours               teamd
774c9687f4c8   docker-fpm-frr:latest                "/usr/bin/docker_ini…"   23 hours ago         Up 23 hours               bgp
8748e8154d31   docker-lldp:latest                   "/usr/bin/docker-lld…"   23 hours ago         Up 23 hours               lldp
0626f850a892   docker-platform-monitor:latest       "/usr/bin/docker_ini…"   23 hours ago         Up 23 hours               pmon
7d3eaaf7e2a0   docker-router-advertiser:latest      "/usr/bin/docker-ini…"   23 hours ago         Up 23 hours               radv
facdfb328b7d   docker-syncd-brcm:latest             "/usr/local/bin/supe…"   23 hours ago         Up 23 hours               syncd
09845b59068c   docker-orchagent:latest              "/usr/bin/docker-ini…"   23 hours ago         Up 23 hours               swss
ad1f6154685c   docker-eventd:latest                 "/usr/local/bin/supe…"   23 hours ago         Up 23 hours               eventd
db6e0fdce9a6   docker-database:latest               "/usr/local/bin/dock…"   23 hours ago         Up 23 hours               database
admin@sonic:~$ curl http://localhost:9102
<html>
                <head><title>SONiC Exporter</title></head>
                <body>
                <h1>SONiC Exporter</h1>
                <p><a href="/metrics">Metrics</a></p>
                </body>
                </html>admin@sonic:~$

admin@sonic:~$ curl http://localhost:9102/metrics
curl: (52) Empty reply from server

admin@sonic:~$ sudo docker logs -f sonic_exporter
level=info msg="Starting up" exporter-git-hash=e1d7f32 exporter-version=0.2.0 lib-node-version=v1.3.1 lib-prometheus-version=v1.11.0
level=info msg="Configuration loaded" port=1234 vrf=
level=info msg="node-library: Parsed flag --collector.filesystem.mount-points-exclude" flag="^/(dev|proc|run/credentials/.+|sys|var/lib/docker/.+)($|/)"
level=info msg="node-library: Parsed flag --collector.filesystem.fs-types-exclude" flag="^(autofs|binfmt_misc|bpf|cgroup2?|configfs|debugfs|devpts|devtmpfs|fusectl|hugetlbfs|iso9660|mqueue|nsfs|overlay|proc|procfs|pstore|rpc_pipefs|securityfs|selinuxfs|squashfs|sysfs|tracefs)$"
level=info msg="node-library: Parsed flag --collector.systemd.unit-include" flag=.+
level=info msg="node-library: Parsed flag --collector.systemd.unit-exclude" flag=".+\\.(automount|device|mount|scope|slice)"
level=warning msg="node-library: Unable to get systemd dbus connection, defaulting systemd version to 0" err="dial unix /run/systemd/private: connect: no such file or directory"
level=warning msg="node-library: Detected systemd version is lower than minimum" current=0 minimum=212
level=warning msg="node-library: Some systemd state and timer metrics will not be available"
level=info msg="SONiC Prometheus exporter running"
level=info msg="Configuration change detected, restarting exporter..."
level=info msg="Starting up" exporter-git-hash=e1d7f32 exporter-version=0.2.0 lib-node-version=v1.3.1 lib-prometheus-version=v1.11.0
level=info msg="Configuration loaded" port=9102 vrf=
level=info msg="node-library: Parsed flag --collector.filesystem.mount-points-exclude" flag="^/(dev|proc|run/credentials/.+|sys|var/lib/docker/.+)($|/)"
level=info msg="node-library: Parsed flag --collector.filesystem.fs-types-exclude" flag="^(autofs|binfmt_misc|bpf|cgroup2?|configfs|debugfs|devpts|devtmpfs|fusectl|hugetlbfs|iso9660|mqueue|nsfs|overlay|proc|procfs|pstore|rpc_pipefs|securityfs|selinuxfs|squashfs|sysfs|tracefs)$"
level=info msg="node-library: Parsed flag --collector.systemd.unit-include" flag=.+
level=info msg="node-library: Parsed flag --collector.systemd.unit-exclude" flag=".+\\.(automount|device|mount|scope|slice)"
level=warning msg="node-library: Unable to get systemd dbus connection, defaulting systemd version to 0" err="dial unix /run/systemd/private: connect: no such file or directory"
level=warning msg="node-library: Detected systemd version is lower than minimum" current=0 minimum=212
level=warning msg="node-library: Some systemd state and timer metrics will not be available"
level=info msg="SONiC Prometheus exporter running"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x2c pc=0x91c1e2]

goroutine 61 [running]:
github.com/sirupsen/logrus.(*Logger).Logf(0xb4671b, 0x7, {0xb49e2d, 0x0}, {0xc000308f80, 0x0, 0x0})
        /go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/logger.go:149 +0x22
github.com/sirupsen/logrus.(*Logger).Errorf(...)
        /go/pkg/mod/github.com/sirupsen/logrus@v1.8.1/logger.go:183
main.NodeCollector.Collect.func1({0xb4671b, 0x7}, {0x7f8559619f38, 0xc0001dc2d0})
        /build/node_collector.go:62 +0xf1
created by main.NodeCollector.Collect
        /build/node_collector.go:59 +0xb5

admin@sonic:~$ show ver

SONiC Software Version: SONiC.202211.181693-c34e3ff86
Distribution: Debian 11.5
Kernel: 5.10.0-18-2-amd64
Build commit: c34e3ff86
Build date: Tue Nov 29 19:21:22 UTC 2022
Built by: AzDevOps@vmss-soni0001XK

```

**Comment by rokernel** [comment]
I also confirm `node_exporter-1.5.0.linux-amd64` works fine.

**Comment by bluecmd** [comment]
Hi! I have fixed this issue now, try with 0.3.0!
